src/CodeGen_X86.cpp

```diff
@@ -277,13 +248,147 @@ void CodeGen_X86::visit(const Sub *op) {
     }
 }
 
+void CodeGen_X86::visit(const GT *op) {
+    Type t = op->a.type();
+    int bits = t.width * t.bits;
+    if (t.width == 1 || bits % 128 == 0) {
+        // LLVM is fine for native vector widths or scalars
+        CodeGen_Posix::visit(op);
+    } else {
+        // Non-native vector widths get legalized poorly by llvm. We
+        // split it up ourselves.
+        Value *a = codegen(op->a), *b = codegen(op->b);
+
+        int slice_size = 128 / t.bits;
+
+        vector<Value *> result;
+        for (int i = 0; i < op->type.width; i += slice_size) {
+            Value *sa = slice_vector(a, i, slice_size);
+            Value *sb = slice_vector(b, i, slice_size);
+            Value *slice_value;
+            if (t.is_float()) {
+                slice_value = builder->CreateFCmpOGT(sa, sb);
+            } else if (t.is_int()) {
+                slice_value = builder->CreateICmpSGT(sa, sb);
+            } else {
+                slice_value = builder->CreateICmpUGT(sa, sb);
+            }
+            result.push_back(slice_value);
+        }
+
+        value = concat_vectors(result);
+        value = slice_vector(value, 0, t.width);
+    }
+}
+
+void CodeGen_X86::visit(const EQ *op) {
+    Type t = op->a.type();
+    int bits = t.width * t.bits;
+    if (t.width == 1 || bits % 128 == 0) {
+        // LLVM is fine for native vector widths or scalars
+        CodeGen::visit(op);
+    } else {
+        // Non-native vector widths get legalized poorly by llvm. We
+        // split it up ourselves.
+        Value *a = codegen(op->a), *b = codegen(op->b);
+
+        int slice_size = 128 / t.bits;
```

Comment:
I'd like 3 x float vectors (for example) to go to 128 bits, not 256.

On Wed, Jan 7, 2015 at 4:32 PM, Dillon Sharlet <notifications@github.com>
wrote:

> In src/CodeGen_X86.cpp
> <https://github.com/halide/Halide/pull/615#discussion-diff-22628118>:
>
> > +        value = slice_vector(value, 0, t.width);
> > +    }
> > +}
> > +
> > +void CodeGen_X86::visit(const EQ *op) {
> > +    Type t = op->a.type();
> > +    int bits = t.width * t.bits;
> > +    if (t.width == 1 || bits % 128 == 0) {
> > +        // LLVM is fine for native vector widths or scalars
> > +        CodeGen::visit(op);
> > +    } else {
> > +        // Non-native vector widths get legalized poorly by llvm. We
> > +        // split it up ourselves.
> > +        Value *a = codegen(op->a), *b = codegen(op->b);
> > +
> > +        int slice_size = 128 / t.bits;
>
> How about:
>
> int slice_size = native_vector_bits() / t.bits;
>
> ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/615/files#r22628118>.
>